include/mumble/Base64.hpp

```diff
@@ -3,31 +3,175 @@
+
 #ifndef MUMBLE_BASE64_HPP
 #define MUMBLE_BASE64_HPP
 
-#include "Macros.hpp"
-#include "NonCopyable.hpp"
 #include "Types.hpp"
 
-#include <memory>
+#include <array>
+#include <cmath>
+#include <limits>
+#include <stdexcept>
+
+#include <gsl/span>
 
 namespace mumble {
-class MUMBLE_EXPORT Base64 : NonCopyable {
-public:
-	class P;
+namespace base64 {
+	using StrView      = gsl::span< char >;
+	using StrViewConst = gsl::span< const char >;
+
+	static constexpr char PAD_CHAR = '=';
+
+	static constexpr std::byte operator+(const std::byte lhs, const std::byte rhs) {
+		const auto ret = std::to_integer< uint8_t >(lhs) + std::to_integer< uint8_t >(rhs);
+		return std::byte(ret);
+	}
+
+	// The standard doesn't overload this operator with both operands as std::byte...
+	static constexpr std::byte operator<<(const std::byte lhs, const std::byte rhs) {
+		return lhs << std::to_integer< uint8_t >(rhs);
+	}
+
+	static constexpr std::byte charToByte(const char ch) {
+		if (ch >= 'A' && ch <= 'Z')
+			return std::byte(ch - 'A');
+		if (ch >= 'a' && ch <= 'z')
+			return std::byte(ch - 'a' + ('Z' - 'A') + 1);
+		if (ch >= '0' && ch <= '9')
+			return std::byte(ch - '0' + ('Z' - 'A') + ('z' - 'a') + 2);
+		if (ch == '+' || ch == '-')
+			return std::byte(62);
+		if (ch == '/' || ch == '_')
+			return std::byte(63);
+
+		throw std::range_error(std::string() + ch + " is not a valid character!");
+	}
+
+	static constexpr char byteToChar(const std::byte byte) {
+		const auto num = std::to_integer< char >(byte);
+
+		if (num <= 25)
+			return num + 'A';
+		if (num >= 26 && num <= 51)
+			return num + 'a' - ('Z' - 'A') - 1;
+		if (num >= 52 && num <= 61)
+			return num + '0' - ('Z' - 'A') - ('z' - 'a') - 2;
+		if (num == 62)
+			return '+';
+		if (num == 63)
+			return '/';
+
+		throw std::range_error(std::string() + num + " is not a valid byte!");
+	}
+
+	static inline bool isValid(const char ch) {
+		try {
+			charToByte(ch);
+		} catch (const std::range_error &) {
+			return false;
+		}
+
+		return true;
+	}
+
+	static inline bool isValid(const StrViewConst in) {
+		for (const auto ch : in) {
+			if (!isValid(ch)) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	static constexpr size_t decodedSize(const StrViewConst in) {
+		size_t size = (in.size() / 4) * 3;
+		// Base64 can have up to 2 pad characters, in order to make the string a multiple of 4.
+		// In order to calculate the exact output size we have to ignore padding.
+		if (auto iter = in.rbegin(); *iter == PAD_CHAR) {
+			size -= *++iter == PAD_CHAR ? 2U : 1U;
+		}
+
+		return size;
+	}
+
+	static constexpr size_t encodedSize(const BufViewConst in) {
+		// ceil((float) a / 3) == (a + 2) / 3, assuming a is an integer.
+		return ((in.size() + 2) / 3) * 4;
+	}
+
+	static constexpr size_t decode(BufView out, StrViewConst in) {
+		if (in.empty()) {
+			return 0;
+		}
+
+		if (in.size() % 4) {
```

Comment:
```suggestion
		if (in.size() % 4 != 0) {
```
makes it more obvious to read